core/src/duplicate_repair_status.rs

```diff
@@ -239,20 +239,69 @@ impl AncestorRequestStatus {
+                ) {
+                    (
+                        Some((mismatch_i, DuplicateAncestorDecision::EarliestMismatchFound(_))),
+                        true,
+                    ) => {
+                        // It's impossible have a different version of an earlier ancestor, but
+                        // then also have the same version of a later ancestor.
+                        let (mismatch_slot, mismatch_agreed_upon_hash) =
+                            agreed_response[*mismatch_i];
+                        let mismatch_our_frozen_hash = blockstore.get_bank_hash(mismatch_slot);
+                        info!(
+                            "When processing the ancestor sample for {}, there was a mismatch
+                            for {mismatch_slot}: we had frozen hash {:?} and the cluster agreed upon
+                            {mismatch_agreed_upon_hash}. However for a later ancestor {ancestor_slot}
+                            we have agreement on {our_frozen_hash} as the bank hash. This should never
+                            be possible, something is wrong or the cluster sample is invalid.
+                            Rejecting and queuing the ancestor hashes request for retry",
+                            self.requested_mismatched_slot,
+                            mismatch_our_frozen_hash
+                        );
+                        return DuplicateAncestorDecision::InvalidSample;
+                    }
+                    (
+                        Some((mismatch_i, DuplicateAncestorDecision::EarliestAncestorNotFrozen(_))),
+                        true,
+                    ) => {
+                        // In this case an earlier ancestor was not frozen in our blockstore,
+                        // however this later ancestor is matching with our version. This most
+                        // likely should never happen however it could happen if we initiate an
+                        // ancestor_hashes request immediately after startup from snapshot, we will
+                        // have a match for the snapshot bank, however we don't have any of the
+                        // ancestors frozen
+                        let (mismatch_slot, mismatch_agreed_upon_hash) =
+                            agreed_response[*mismatch_i];
+                        info!(
+                            "When processing the ancestor sample for {}, an earlier ancestor {mismatch_slot}
+                            was agreed upon by the cluster with hash {mismatch_agreed_upon_hash} but not
+                            frozen in our blockstore. However for a later ancestor {ancestor_slot} we have
+                            agreement on {our_frozen_hash} as the bank hash. This should only be possible if
+                            we have just started from snapshot and immediately encountered a duplicate block,
+                            otherwise something is seriously wrong. Continuing with the repair",
```

Comment:
"Continuing with the repair" rn will still repair everything from the first missing block. However we could go even further and detect the snapshot root and only repair things after. Probably no point in repairing before.

Reply:
I think this is fine for now